subsys/net/lib/conn_mgr/conn_mgr_connectivity.h

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief TODO
+ */
+
+#ifndef ZEPHYR_INCLUDE_CONN_MGR_CONNECTIVITY_H_
+#define ZEPHYR_INCLUDE_CONN_MGR_CONNECTIVITY_H_
+
+#include <zephyr/device.h>
+#include <zephyr/net/net_if.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief TODO: DESCRIBE
+ * @defgroup conn_mgr_connectivity TODO: DESCRIBE
+ * @ingroup networking TODO: GROUP
+ * @{
+ */
+
+/**
+ * @brief Network L2 connectivity API structure
+ *
+ * Used to provide an interface to connect parameters and procedures
+ */
+struct conn_mgr_binding;
+struct conn_mgr_conn_api {
+	/**
+	 * @brief When called, the L2 connectivity implementation should start attempting to
+	 * establish connectivity for the bound iface pointed to by if_conn->iface.
+	 *
+	 * Must be non-blocking.
+	 *
+	 * Called by conn_mgr_if_connect.
+	 */
+	int (*connect)(const struct conn_mgr_binding *binding);
+
+	/**
+	 * @brief When called, the L2 connectivity implementation should stop disconnect and stop
+	 * any in-progress attempts to establish connectivity for the bound iface pointed to by
+	 * if_conn->iface.
+	 *
+	 * Must be non-blocking.
+	 *
+	 * Called by conn_mgr_if_disconnect.
+	 */
+	int (*disconnect)(const struct conn_mgr_binding *binding);
+
+	/**
+	 * @brief Called once for each iface that has been bound to this L2
+	 * connectivity implementation.
+	 *
+	 * L2 connectivity implementations should use this callback to perform any required
+	 * per-bound-iface initialization.
+	 *
+	 * TODO: When is this now called?
```

Comment:
We have a `conn_mgr_init()`  initialization function called at boot (application-level priority). This in order calls per-interface `conn_mgr_init_cb()`. I think the binding initialization should take place in there.

Apart from initializing the binding, we should also check if the autoconnect feature is set on an interface. Interfaces should be already in admin UP at this point (unless those configured not to be brought up on boot), and we've already missed an event for that (as no event handler was set at that point).

Reply:
Should the auto-connect behavior be in this PR, or in the next PR where we plan to re-write conn_mgr's connection logic? I had been thinking the latter rather than the former

I'll introduce the flags field at least so that it is earmarked

Reply:
As for me, we could do it in a separate PR to keep this one easier to review. We should cover the binding initialization though.